### PR TITLE
Add `mypy` type check to `annotation/`

### PR DESCRIPTION
**Pull request by Jiaqi-Lv**
- [x] dsl.py
- [x] storage.py

Had to edit some files in `utils/` and and `tools/` because they were failing mypy.
Also added a few additional pytests.

### CODE REVIEW

File: .github/workflows/mypy-type-check.yml [hunk 1]
```diff
@@ -39,5 +39,6 @@ jobs:
           tiatoolbox/__main__.py \
           tiatoolbox/typing.py \
           tiatoolbox/tiatoolbox.py \
-          tiatoolbox/utils/ \
-          tiatoolbox/tools/
+          tiatoolbox/utils/*.py \
```
Comment (shaneahmed) [inline .github/workflows/mypy-type-check.yml:42]: using `*.py` excludes files in sub folders for some reason.

Comment (Jiaqi-Lv) [reply .github/workflows/mypy-type-check.yml]: Oh Ok. I think we will just specify the directory then, mypy will recursively check the directory

File: tiatoolbox/annotation/storage.py [hunk 2]
```diff
@@ -425,10 +427,12 @@ def decode_polygon(offset: int = 0) -> tuple[list[np.ndarray], int]:
         raise ValueError(msg)
 
 
-class AnnotationStore(ABC, MutableMapping):
+class AnnotationStore(ABC, MutableMapping[str, Annotation]):
     """Annotation store abstract base class."""
 
-    def __new__(cls: ABC, *args: str, **kwargs: int) -> ABC:  # noqa: ARG003
+    A = TypeVar("A", bound="AnnotationStore")
```
Comment (shaneahmed) [inline tiatoolbox/annotation/storage.py:441]: Maybe remove this or rename to a more meaningful name.

**Review by shaneahmed — APPROVED** [review]
Thanks @Jiaqi-Lv

**Pull request merged**